gluten-core/src/main/java/org/apache/gluten/memory/memtarget/TreeMemoryTargets.java

```diff
@@ -118,7 +125,29 @@ public long borrow(long size) {
         return 0;
       }
       ensureFreeCapacity(size);
-      return borrow0(Math.min(freeBytes(), size));
+      long requiredSize = Math.min(freeBytes(), size);
+      long granted = borrow0(requiredSize);
+
+      // If isDynamicCapacity is true, which means it is controlled by vanilla Spark,
+      // and if the granted memory is less than the required size, it may be because
+      // this task holds memory exceeding maxMemoryPerTask. Therefore, we actively retry
+      // spilling all memory. After this, if there is still not enough memory acquired,
+      // it should result in an OOM.
+      if (granted < requiredSize && isDynamicCapacity) {
+        LOGGER.info(
+            "Exceed Spark perTaskLimit with maxTaskSizeDynamic when "
+                + "require:{} got:{}, try spill all.",
+            requiredSize,
+            granted);
+        long spilled = TreeMemoryTargets.spillTree(this, Long.MAX_VALUE);
+        long remain = requiredSize - granted;
+        if (spilled > remain) {
+          granted += borrow0(remain);
+        } else {
+          // OOM
+        }
+      }
+      return granted;
```

Comment:
hi @kecookier 

Can we add a individual memory target for doing theretries?

Something like

```java
  public static class RetryOnOomMemoryTarget extends MemoryTarget {
    private final TreeMemoryTarget target;

    @Override
    public long borrow(long size) {
      final long granted = target.borrow(size);
      if (granted >= size) {
        return granted;
      }
      // Granted < size. Spill the underlying memory target then retry borrowing.
      final long remaining = size - granted;
      TreeMemoryTargets.spillTree(target, Long.MAX_VALUE);
      final long granted2 = target.borrow(remaining);
      return granted + granted2;
    }
```

I can help do a further benchmark too see if chaining another memory target hurts the overall performance. Thought we can make a try like this anyway.

Reply:
@zhztheplayer  I see, something like ThrowOnOomMemoryTarget( RetryOnOomMemoryTarget( Overacquired(...) ) )?

Reply:
Something like that but I assume we tended to add it to [this line](https://github.com/apache/incubator-gluten/blob/05b1e7a28fa8b2915b573496466cbe94c7baab9b/gluten-core/src/main/java/org/apache/gluten/memory/memtarget/MemoryTargets.java#L64)? Since shared mode is only interested.

Reply:
We can add a comment like "Retry spilling is needed in shared mode because ... " or something

```java
    if (GlutenConfig.getConf().memoryIsolation()) {
      return TreeMemoryConsumers.isolated()
          .newConsumer(tmm, name, spiller, virtualChildren);
    } else {
      // Retry spilling is needed in shared mode because...
      return MemoryTargets.retrySpillOnOom(
          TreeMemoryConsumers.shared()
              .newConsumer(tmm, name, spiller, virtualChildren));
    }
```

Reply:
@zhztheplayer  Thanks, I'll update the code soon.